### PR TITLE
kubectl create ingress command

### LINKED ISSUE #43667: kubectl create ingress / kubectl expose
I think it would be very useful to be able to expose services via an ingress with a single command.  In particular, because ingress seems to always need a hostname, and that will always be specific to the installation.

There are two ways that immediately spring to mind to do this, and I'd appreciate some input on which is preferable before I start coding ( cc @kubernetes/sig-cli-feature-requests ) :

1. `kubectl expose service myapp  --ingress myapp.example.com --tls-acme`
1. `kubectl create ingress myapp --host myapp.example.com --tls-acme`

The second would rely on a service existing with the same name and would rely on the service port being port 80, whereas the first performs more error checking and is less magical.  But I can also appreciate how the second continues to nudge people towards a well-established pattern.

cc @aledbf for ingress

### PR DESCRIPTION
**Pull request by justinsb**
`kubectl create ingress` is a subcommand of `kubectl create` which creates
an ingress.

It takes a name positional argument, which is the name of the ingress to
create.

It also takes a repeated --host parameter, which is the hostname to use
in the ingress.

If --service-name or --service-port are specified, those values will be
used for the backend service name and port.  Otherwise we default the
name to the same as the ingress name, and the port to 80 (because
ingress is primarily an HTTP concept).

If --tls-acme is specified, TLS will also be configured on the ingress,
with the annotations as used by the kube-lego LetsEncrypt controller,
and a secret named `tls-<ingress name>`, configured for all the
specified hosts.

Close #43667

```release-note
`kubectl create ingress` subcommand added
```

### CODE REVIEW

File: pkg/kubectl/cmd/create_ingress.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		},
+	}
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.IngressV1Beta1GeneratorName)
+	cmd.Flags().StringSlice("host", []string{}, "Host name for the ingress record")
+	cmd.Flags().Bool("tls-acme", false, "Enables ACME (LetEncrypt) support for automatic TLS")
+	cmd.Flags().String("service-name", "", "Name of backend service; defaults to same as ingress name.")
+	cmd.Flags().String("service-port", "80", "Port of backend service.")
```
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/create_ingress.go]: Give a explicit tip of default service-port for users? Such as `Port of backend service. Default to 80.`

File: pkg/kubectl/ingress.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+			ingress.Annotations = make(map[string]string)
+		}
+		ingress.Annotations[ingressAnnotationNameTLSAcme] = "true"
+	}
+
+	return &ingress, nil
+}
+
+// validate validates required fields are set to support structured generation
+func (g *IngressV1Beta1) validate() error {
```
Comment (xiangpengzhao) [inline pkg/kubectl/ingress.go]: If `service-port` is specified, should we validate if the `service-port` has a valid name or be in a valid range?

File: pkg/kubectl/ingress.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+			delegate.ServicePort = intstr.FromString(servicePortString)
+		} else {
+			return nil, fmt.Errorf("expected string or int, saw %v for 'service-port'", params["service-port"])
+		}
+	}
+
+	return delegate.StructuredGenerate()
+}
+
+// StructuredGenerate outputs a ResourceQuota object using the configured fields
```
Comment (xiangpengzhao) [inline pkg/kubectl/ingress.go]: s/a ResourceQuota/an Ingress

**Review by xiangpengzhao — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/create_ingress.go (hunk 1 above)
Comment (justinsb) [inline pkg/kubectl/cmd/create_ingress.go]: Done!

File: pkg/kubectl/ingress.go (hunk 2 above)
Comment (justinsb) [inline pkg/kubectl/ingress.go]: We would normally hit k8s API validation, anyway, but I added it

File: pkg/kubectl/ingress.go (hunk 3 above)
Comment (justinsb) [inline pkg/kubectl/ingress.go]: Thanks - nice spot

**Comment by justinsb** [comment]
Made fixes per review

File: pkg/kubectl/ingress.go (hunk 3 above)
Comment (xiangpengzhao) [inline pkg/kubectl/ingress.go]: :)

**Comment by justinsb** [comment]
@fabianofranz ping for /approve

**Comment by fabianofranz** [comment]
@kubernetes/sig-cli-misc

File: pkg/kubectl/cmd/create_ingress.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			err := CreateIngress(f, cmdOut, cmd, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.IngressV1Beta1GeneratorName)
+	cmd.Flags().StringSlice("host", []string{}, "Host name for the ingress record")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/create_ingress.go]: Looks like this flag is required, so it needs `cmd.MarkFlagRequired`.

File: pkg/kubectl/cmd/create_ingress.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	cmd.Flags().Bool("tls-acme", false, "Enables ACME (LetEncrypt) support for automatic TLS")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/create_ingress.go]: Let’s Encrypt

**Comment by fabianofranz** [comment]
@justinsb Sorry about the delay! Just a couple minor comments, make sure you squash then this is good to go.

**Comment by justinsb** [comment]
Thanks @fabianofranz - should be all fixed up & squashed

File: pkg/kubectl/cmd/create_ingress.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/util/i18n"
+)
+
+var (
+	ingressLong = templates.LongDesc(i18n.T(`
+    Create an ingress with the specified name.`))
+
+	ingressExample = templates.Examples(i18n.T(`
+    # Create a new ingress named my-ingress supporting TLS with LetsEncrypt
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/create_ingress.go]: Let’s Encrypt

**Comment by fabianofranz** [comment]
> should be all fixed up & squashed

not yet, forgot to `git push`?  ;)

**Comment by justinsb** [comment]
@k8s-bot bazel test this

Weird git error

**Comment by justinsb** [comment]
Thanks @fabianofranz - I had indeed failed to push!  Should be fixed now!

**Comment by fabianofranz** [comment]
LGTM from a CLI perspective, @smarterclayton mind taking a conceptual look at this?

**Comment by justinsb** [comment]
ping @smarterclayton

**Comment by justinsb** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by justinsb** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by grodrigues3** [comment]
@justinsb this looks like a feature that was not lgtm'd prior to code freeze.  Removing from the milestone.

**Comment by justinsb** [comment]
ping @smarterclayton

File: pkg/kubectl/cmd/create_ingress_test.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func (t *testIngressPrinter) HandledResources() []string {
+	return []string{}
+}
+
+func TestCreateIngress(t *testing.T) {
+	f, tf, _, _ := cmdtesting.NewAPIFactory()
+	printer := &testIngressPrinter{}
+	tf.Printer = printer
```
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/create_ingress_test.go]:
```
I0613 06:12:09.973] pkg/kubectl/cmd/create_ingress_test.go:53: cannot use printer (type *testIngressPrinter) as type printers.ResourcePrinter in assignment:
I0613 06:12:09.974] 	*testIngressPrinter does not implement printers.ResourcePrinter (missing IsGeneric method)
```

**Review by xiangpengzhao — COMMENTED** [review]
`pull-kubernetes-bazel` failed.

**Comment by justinsb** [comment]
@smarterclayton appears to have turned off all github notifications.

/unassign smarterclayton

**Comment by smith** [comment]
I would like to shadow this code review

**Comment by liggitt** [comment]
given the [in progress move to networking.k8s.io](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190125-ingress-api-group.md), and the impending removal of extensions/v1beta1, let's avoid adding new uses of extensions/v1beta1, especially in clients like kubectl that have longer skew support

**Comment by liggitt** [comment]
closing this for now, can revisit once ingress v1 plans are completed in 1.15

/closew

**Pull request closed without merging**